nifi-nar-bundles/nifi-slack-bundle/nifi-slack-processors/src/main/java/org/apache/nifi/processors/slack/FetchSlack.java

```diff
@@ -0,0 +1,295 @@
+    try {
+      ArrayList<FlowFile> successfulResponses = new ArrayList<>();
+      for (JsonObject file : filesToDownload) {
+        String url = file.getString("url_private_download");
+
+        HttpUriRequest request = RequestBuilder.get().setUri(url).build();
+        HttpResponse response = client.execute(request);
+        int statusCode = response.getStatusLine().getStatusCode();
+
+        if (inStatusCodeFamily(200, statusCode)) {
+          FlowFile responseFlowFile = session.create(requestFlowFile);
+          session.write(responseFlowFile, out -> response.getEntity().writeTo(out));
+
+          Map<String, String> attributes = fillAttributes(requestFlowFile.getAttributes(), file, response);
+          responseFlowFile = session.putAllAttributes(responseFlowFile, attributes);
+
+          session.getProvenanceReporter().fetch(responseFlowFile, url);
+          successfulResponses.add(responseFlowFile);
+        } else if (inStatusCodeFamily(100, statusCode)
+          || inStatusCodeFamily(300, statusCode)
+          || inStatusCodeFamily(400, statusCode)) {
+          session.transfer(requestFlowFile, REL_NO_RETRY);
+          break;
+        } else if (inStatusCodeFamily(500, statusCode)) {
+          session.transfer(requestFlowFile, REL_RETRY);
+          break;
+        }
+      }
+
+      if (successfulResponses.size() == filesToDownload.size()) {
+        session.transfer(successfulResponses, REL_RESPONSE);
+        session.transfer(requestFlowFile, REL_SUCCESS_REQ);
+      }
+    } catch (IOException e) {
+      session.transfer(requestFlowFile, REL_FAILURE);
+    }
+  }
+
+  private List<JsonObject> collectFilesToDownload(InputStream requestInputStream) {
+    List<JsonObject> filesToDownload = new ArrayList<>();
+    try (JsonReader reader = Json.createReader(requestInputStream)) {
+      JsonObject jsonObject = reader.readObject();
+      JsonArray files = jsonObject.getJsonArray("files");
+      if (files != null) {
+        for (JsonValue file : files) {
+          String url = ((JsonObject) file).getString("url_private_download");
+          if (url != null) {
+            filesToDownload.add((JsonObject) file);
+          }
+        }
+      }
+    }
+
+    return filesToDownload;
+  }
+
+  private HashMap<String, String> fillAttributes(Map<String, String> attributes, JsonObject file, HttpResponse response) {
+    HashMap<String, String> map = new HashMap<>(attributes);
+    String value = response.getFirstHeader("Content-Disposition").getValue();
+    Matcher matcher = filenamePattern.matcher(value);
+    if (matcher.matches()) {
+      map.put("response.filename", matcher.group(1));
+    }
+
+    fromheader(map, response, "Content-Length", "response.content-length");
+    fromheader(map, response, "Content-type", "response.content-type");
+    fromheader(map, response, "Date", "response.date");
+
+    fromJson(map, file, "name", "message.name");
```

Comment:
Are these 'message' prefix correct? I wonder if it would be better we call them as 'slack.file.name' ... etc